lib/classes/Variables.js

```diff
@@ -145,7 +148,7 @@ class Variables {
   }
 
   getValueFromSelf(variableString) {
-    let valueToPopulate = _.cloneDeep(this.service);
+    let valueToPopulate = this.service;
```

Comment:
[sorry for the novel...]

TL;DR: 1) avoid infinite recursion 2) serverless depends on in-place affect on the original service object.

Detailed Discussion:

So, my reasoning was at least correctish...  Not sure that my "evidence ... [of] expansion on demand" was properly interpreted.

Re-adding the `_.deepClone` causes an infinite recursion (technically, it only got 177 iterations deep before I paused execution).  It copies the `custom.retailStream: ${self:}` attribute and value pair that is to be replaced and then (because `if (property !== this.service)` fails to identify the property is a copy of the service [on line 73]) attempts to `populateProperty` on the service copy (that still contains the `${self:}`.  This repeats.  That's layer one of the infinite recursion.

I could write a rough equivalency check to avoid the recursion on line 73.  Something like `if (!(property.service && property.service === this.service.service))` would probably suffice in most cases but `_.deepEquals(property, this.service)` definitely wouldn't work since the two will have already diverged (the parent service has an extra level that the cloned service does not).

The rough equivalency check will get us past the recursion on line 73 but then the traverse on line 41 will attempt to traverse the `${self:}` and as a result get another one, et cetera, to "infinity and beyond!". 😄 

It seems to me that if we want to protect the "original" in-memory copy of the service then we ought to do that once at the outset (i.e. in the constructor or the like).  The problem is that this doesn't cover the bare use of `populateProperty(<something-not-the-service>)` but a revision to the call that adds a "called from populateService" parameter could distinguish between the cases.  Passing the property in the `populateProperty` case but (relying on instance state and) not passing the service in the `populateService` case is a bit of a mental model mix-up that is legacy now.

Hmm... (sorry, writing as I code and test...)

It turns out that the framework is actually **relying** on the revision of the original service object.  Adding the use of `_.cloneDeep` to the constructor causes the copy of the service that the framework uses to take action not on the altered copy but the original and thus I observed:
```
  Serverless Error ---------------------------------------
 
     Bucket name must start with a letter or number. ${self:custom.helloRetail.deploymentBucket}
 
  Stack Trace --------------------------------------------
 
ServerlessError: Bucket name must start with a letter or number. ${self:custom.helloRetail.deploymentBucket}
    at ~/serverless/lib/plugins/aws/lib/validate.js:57:17
From previous event:
    at AwsDeploy.validateS3BucketName (~/serverless/lib/plugins/aws/lib/validate.js:34:8)
    at AwsDeploy.<anonymous> (~/serverless/lib/plugins/aws/deploy/lib/configureStack.js:22:26)
From previous event:
    at AwsDeploy.configureStack (~/serverless/lib/plugins/aws/deploy/lib/configureStack.js:22:10)
From previous event:
    at __dirname.constructor.options.hooks.deploy:initialize (~/serverless/lib/plugins/aws/deploy/index.js:42:10)
    at ~/serverless/lib/classes/PluginManager.js:160:50
    at processImmediate [as _immediateCallback] (timers.js:383:17)
From previous event:
    at PluginManager.run (~/serverless/lib/classes/PluginManager.js:160:22)
    at Serverless.run (~/serverless/lib/Serverless.js:91:31)
    at ~/serverless/bin/serverless:23:50
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
 
  Your Environment Information -----------------------------
     OS:                 darwin
     Node Version:       4.6.2
     Serverless Version: 1.5.1
```

So therefore... I can remove the top level clone and still do a clone below if the caller doesn't ask for an in-place population (which populate service would ask for).  The remaining potential concern space then is that the service could be improperly altered by the population of a property, via a `${self:<var>*}` declaration that isn't a part of the service.  However, by definition, the `self` variable type only references the service, so that concern space is empty/size zero.

Maybe I fail to understand the case that you were concerned about that motivated the cloning?  Sorry if so.  I'd love some illumination.